### PR TITLE
Reorganize GMX Setup Utils

### PR DESCRIPTION
**Pull request by kanduri**

### CODE REVIEW

File: src/gromacs/nblib/gmxsetup.h [hunk 1]
```diff
@@ -67,17 +67,32 @@ enum class CombinationRule : int
...
+    void setParticlesOnGrid(std::unique_ptr<nonbonded_verlet_t>& nbv);
 
-    std::unique_ptr<GmxForceCalculator> setupGmxForceCalculator();
+    //! Sets Particle Types and Charges and VdW params
+    void setAtomProperties(std::unique_ptr<nonbonded_verlet_t>& nbv);
 
+    //! Sets up non-bonded verlet on the GmxFOrceCalculator
+    std::unique_ptr<nonbonded_verlet_t> setupNbnxmInstance();
+
+public:
```
Comment (victorusu) [inline src/gromacs/nblib/gmxsetup.h]: These data members are not being used outside the class.
So, I vote that they actually disappear and get passed as arguments to the functions that need them...

**Review by victorusu — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/gmxsetup.h [hunk 2]
```diff
@@ -67,17 +67,32 @@ enum class CombinationRule : int
...
+    //! Sets up and returns a GmxForceCalculator
+    std::unique_ptr<GmxForceCalculator> setupGmxForceCalculator();
+
+private:
+    //! Reads Topology data from the user-facing layer to translate to GMX data structures
     void unpackTopologyToGmx();
 
-    std::unique_ptr<nonbonded_verlet_t> setupNbnxmInstance();
+    //! Puts particles on a grid based on bounds specified by the box
+    void setParticlesOnGrid(std::unique_ptr<nonbonded_verlet_t>& nbv);
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.h]: this should take the coordinates as the argument, not the nbv which is the object being built and owned by
NbvSetupUtil

File: src/gromacs/nblib/gmxsetup.h [hunk 3]
```diff
@@ -67,17 +67,32 @@ enum class CombinationRule : int
...
 
-    std::unique_ptr<GmxForceCalculator> setupGmxForceCalculator();
+    //! Sets Particle Types and Charges and VdW params
+    void setAtomProperties(std::unique_ptr<nonbonded_verlet_t>& nbv);
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.h]: Same here, this needs to have the nbf parameters as arguments.

File: src/gromacs/nblib/tests/nbnxnsetup.cpp [hunk 4]
```diff
@@ -48,17 +48,23 @@
...
+#include "testsystems.h"
+
 namespace nblib
 {
 
 namespace test
 {
-/*
+
 TEST(NBlibTest, NbfpSetCorrectly)
```
Comment (ejjordan) [inline src/gromacs/nblib/tests/nbnxnsetup.cpp]: This should be CanConstructNbvSetupUtil and should test that construction does not throw.

File: src/gromacs/nblib/tests/nbnxnsetup.cpp [hunk 5]
```diff
@@ -48,17 +48,23 @@
...
-/*
-TEST(NBlibTest, NbfpSetCorrectly)
-{
 
+TEST(NBlibTest, CanConstructNbvSetupUtil)
+{
+    ArgonSimulationStateBuilder argonSimulationStateBuilder;
+    SimulationState             system = argonSimulationStateBuilder.setupSimulationState();
+    NBKernelOptions             options;
+    EXPECT_NO_THROW(auto nbvSetupUtil = std::make_unique<NbvSetupUtil>(system, options));
```
Comment (ejjordan) [inline src/gromacs/nblib/tests/nbnxnsetup.cpp]: The "auto nbvSetupUtil = " is not needed.

File: src/gromacs/nblib/gmxsetup.h (hunk 1 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: Yes. Working on this.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 6]
```diff
@@ -198,6 +198,31 @@ void NbvSetupUtil::unpackTopologyToGmx()
...
+void NbvSetupUtil::setAtomProperties(std::unique_ptr<nonbonded_verlet_t>& nbv)
+{
+    t_mdatoms mdatoms;
+    // We only use (read) the atom type and charge from mdatoms
+    mdatoms.typeA = const_cast<int*>(system_->topology().getParticleTypeIdOfAllParticles().data());
+    mdatoms.chargeA = const_cast<real*>(system_->topology().getCharges().data());
+    nbv->setAtomProperties(mdatoms, particleInfoAllVdw_);
+}
+
+void NbvSetupUtil::setParticlesOnGrid(std::unique_ptr<nonbonded_verlet_t>& nbv)
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxsetup.cpp]: This function will be needed both here and in the GmxForceCalculator so I think it should be a free function that with parameters nbv, box, coordinates, particleInfoAllVdw.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 7]
```diff
@@ -198,6 +198,31 @@ void NbvSetupUtil::unpackTopologyToGmx()
...
+{
+    matrix box_;
+    gmx::fillLegacyMatrix(system_->box().matrix(), box_);
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxsetup.cpp]: This gets called enough that we should probably just construct this object in box and be able to call box.legacyMatrix().

File: src/gromacs/nblib/gmxsetup.cpp [hunk 8]
```diff
@@ -198,6 +198,31 @@ void NbvSetupUtil::unpackTopologyToGmx()
     }
 }
 
+void NbvSetupUtil::setAtomProperties(std::unique_ptr<nonbonded_verlet_t>& nbv)
+{
+    t_mdatoms mdatoms;
+    // We only use (read) the atom type and charge from mdatoms
+    mdatoms.typeA = const_cast<int*>(system_->topology().getParticleTypeIdOfAllParticles().data());
+    mdatoms.chargeA = const_cast<real*>(system_->topology().getCharges().data());
+    nbv->setAtomProperties(mdatoms, particleInfoAllVdw_);
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxsetup.cpp]: The actual setting in setAtomProperties() should be a separate function that takes the mdatoms constructed in its own function. The mdatoms need to be passed to the GmxForceCalculator anyway so this cannot be a temporary.

**Review by ejjordan — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/gmxsetup.cpp (hunk 7 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.cpp]: Done.

File: src/gromacs/nblib/gmxsetup.cpp (hunk 8 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.cpp]: Made `t_mdatoms` a member of `GmxForceCalculator` and made the created `mdatoms` persistent.

File: src/gromacs/nblib/gmxsetup.cpp (hunk 6 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.cpp]: Done.

File: src/gromacs/nblib/tests/nbnxnsetup.cpp (hunk 5 above)
Comment (kanduri) [reply src/gromacs/nblib/tests/nbnxnsetup.cpp]: Done.

File: src/gromacs/nblib/gmxcalculator.h [hunk 9]
```diff
@@ -106,6 +106,9 @@ class GmxForceCalculator
     matrix box_;
 };
 
+//! Puts particles on a grid based on bounds specified by the box
+void setParticlesOnGrid(SimulationState simulationState, std::unique_ptr<nonbonded_verlet_t>& nbv, std::vector<int>& particleInfoAllVdw_);
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxcalculator.h]: The argument list should be the parameters that are actually used and not the whole simState object so that the data flow is clear.

**Review by ejjordan — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/gmxcalculator.h (hunk 9 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxcalculator.h]: Done.

File: src/gromacs/nblib/box.cpp [hunk 10]
```diff
@@ -72,14 +74,16 @@ Box::Box(real x, real y, real z)
     box_(XX, XX) = x;
     box_(YY, YY) = y;
     box_(ZZ, ZZ) = z;
+
+    gmx::fillLegacyMatrix(box_, legacyMatrix);
 }
 
-Box::Matrix Box::matrix()
+Box::Matrix Box::getMatrix()
```
Comment (ejjordan) [inline src/gromacs/nblib/box.cpp]: We don't use get for getters in gromacs so to stay in line with those coding standards this should stay as matrix().

File: src/gromacs/nblib/box.cpp [hunk 11]
```diff
@@ -72,14 +74,16 @@ Box::Box(real x, real y, real z)
     box_(XX, XX) = x;
     box_(YY, YY) = y;
     box_(ZZ, ZZ) = z;
+
+    gmx::fillLegacyMatrix(box_, legacyMatrix);
```
Comment (ejjordan) [inline src/gromacs/nblib/box.cpp]: You need to add a getter for the legacyMatrix.

File: src/gromacs/nblib/gmxcalculator.cpp [hunk 12]
```diff
@@ -132,7 +133,7 @@ GmxForceCalculator::GmxForceCalculator(SimulationState                        si
     stepWork_(setupStepWorkload(options)),
     enerd_(gmx_enerdata_t(1, 0))
 {
-    gmx::fillLegacyMatrix(simState.box().matrix(), box_);
+    gmx::fillLegacyMatrix(simState.box().getMatrix(), box_);
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxcalculator.cpp]: This should just call the legacyMatrix getter.

**Review by ejjordan — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/box.cpp (hunk 11 above)
Comment (kanduri) [reply src/gromacs/nblib/box.cpp]: The legacy matrix is a C array. So one can return a reference to an array, or copy it in-place. It cannot be returned like STL containers. Hence, I'd fill up such an array in-place and access it when needed. I should add const protections to it.

File: src/gromacs/nblib/box.cpp (hunk 10 above)
Comment (kanduri) [reply src/gromacs/nblib/box.cpp]: The keyword `matrix` as the typedef in GROMACS conflicts with the function name `matrix()`. Hence, had to rename it.

On another note: Does NB-LIB's user facing code have to adhere to GROMACS's naming guidelines? I understand that the translation layer should. Like this case is purely due to conflicts with GROMACS's internal naming.

Comment (ejjordan) [reply src/gromacs/nblib/box.cpp]: There is no collision between matrix and matrix() so the change to getMatrix() at this point would be purely aesthetic. I think we should adhere to gromacs coding style even in the public interface unless there is a compelling reason not to, because then we can just link to that coding style document in our own docs. Also, I find getMatrix() needlessly verbose. Just typing matrix() is clear enough.

File: src/gromacs/nblib/box.cpp (hunk 11 above)
Comment (ejjordan) [reply src/gromacs/nblib/box.cpp]: Fair enough. Since this is the case I think that we should not add anything to the Box class. We will have to call fillLegacyMatrix every time we need to copy anyway so there is no savings in terms of coding or increase in clarity.

Comment (kanduri) [reply src/gromacs/nblib/box.cpp]: Not true. We fill the legacy matrix at the time of construction. Then this legacy matrix is used for any calculations during setup.

File: src/gromacs/nblib/box.cpp (hunk 10 above)
Comment (kanduri) [reply src/gromacs/nblib/box.cpp]: Collision resolved using `using`. Hope the C array usage goes away soon and we use the modern `Matrix` instead.

File: src/gromacs/nblib/gmxcalculator.cpp [hunk 13]
```diff
@@ -146,4 +147,22 @@ gmx::PaddedHostVector<gmx::RVec> GmxForceCalculator::compute()
     return verletForces_;
 }
 
+void setParticlesOnGrid(std::unique_ptr<nonbonded_verlet_t>& nbv,
+                        std::vector<int>&                    particleInfoAllVdw,
+                        const std::vector<gmx::RVec>&        coordinates,
+                        const Box&                           box)
+{
+    const matrix& box_ = box.legacyMatrix;
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxcalculator.cpp]: I think it is not so nice to set box_ in this way. Either this should be a temporary (renamed to box), or box.legacyMatrix should be passed directly functions, or this should be set in its own function (or the builder).

File: src/gromacs/nblib/box.cpp (hunk 11 above)
Comment (ejjordan) [reply src/gromacs/nblib/box.cpp]: Ack

File: src/gromacs/nblib/gmxcalculator.cpp (hunk 13 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxcalculator.cpp]: Renamed the variable. It is a temporary one that dies after that function scope.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 14]
```diff
@@ -246,14 +240,16 @@ std::unique_ptr<GmxForceCalculator> NbvSetupUtil::setupGmxForceCalculator()
...
+    gmxForceCalculator_p->mdatoms_ = setAtomData(system_);
+    setAtomProperties(gmxForceCalculator_p->nbv_, gmxForceCalculator_p->mdatoms_);
 
     // const PairlistSet& pairlistSet = nbv->pairlistSets().pairlistSet(gmx::InteractionLocality::Local);
     // const gmx::index numPairs = pairlistSet.natpair_ljq_ + pairlistSet.natpair_lj_ + pairlistSet.natpair_q_;
     // gmx_cycles_t cycles = gmx_cycles_read();
 
-    matrix box_;
-    gmx::fillLegacyMatrix(system_.box().matrix(), box_);
+    const matrix& box_ = system_.box().legacyMatrix;
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxsetup.cpp]: I would still prefer to have this in the ctor.

File: src/gromacs/nblib/gmxcalculator.h [hunk 15]
```diff
@@ -91,6 +92,9 @@ class GmxForceCalculator
     //! Tasks to perform in an MD Step
     gmx::StepWorkload stepWork_;
 
+    //! Stores atom property data
+    t_mdatoms mdatoms_;
+
```
Comment (sekelle) [inline src/gromacs/nblib/gmxcalculator.h:100]: All of that stuff here needs to be private. The GmxForceCalculator is supposed to be a containment for the
nbv that it manages. Handing out the actual object being managed and then calling other functions on it like setParticlesOnGrid achieves nothing in terms of encapsulation and separation of concerns. The fact that this is not directly a user operable class is absolutely irrelevant to this discussion. The fundamental principle of having each class take care of its own resource, by keeping the actual resource private and manage access through a public interface applies to ALL classes in the design hierarchy, not just the class at the top that interfaces with the user.

File: src/gromacs/nblib/gmxcalculator.h [hunk 16]
```diff
@@ -102,6 +106,12 @@ class GmxForceCalculator
     matrix box_;
 };
 
+//! Puts particles on a grid based on bounds specified by the box
+void setParticlesOnGrid(std::unique_ptr<nonbonded_verlet_t>& nbv,
+                        std::vector<int>&                    particleInfoAllVdw,
+                        const std::vector<gmx::RVec>&        coordinates,
+                        const Box&                           box);
+
```
Comment (sekelle) [inline src/gromacs/nblib/gmxcalculator.h]: This should be a public member function of GmxForceCalculator as discussed and documented in the notes from the hackathon

File: src/gromacs/nblib/gmxsetup.h [hunk 17]
```diff
@@ -64,24 +64,35 @@
...
     std::unique_ptr<GmxForceCalculator> setupGmxForceCalculator();
 
-    SimulationState                  system_;
-    std::shared_ptr<NBKernelOptions> options_;
-
     //! Storage for parameters for short range interactions.
     std::vector<real> nonbondedParameters_;
 
     //! Particle info where all particles are marked to have Van der Waals interactions
     std::vector<int> particleInfoAllVdw_;
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.h]: nonbondedParameters_ and particleInfoAllVdw_ is internal state and needs to be private.
This class still has none of the functions in the public interface that we discussed about, i.e. a function to add
the nonbonded parameters, the atomProperties (VdW, chargest, etc), the AtomTypes etc.
You don't want to pass the entire system and then set all that's needed in an opaque way internally. The necessary data should be fed in piece by piece through a public interface, since this class is supposed to be a builder for GmxForceCalculator.

File: src/gromacs/nblib/integrator.cpp [hunk 18]
```diff
@@ -74,9 +74,7 @@ void LeapFrog::integrate(const real dt)
             x[i][dim] += v[i][dim] * dt;
         }
     }
-    matrix box;
-    gmx::fillLegacyMatrix(simulationState_.box().matrix(), box);
-    put_atoms_in_box(PbcType::Xyz, box, x);
+    put_atoms_in_box(PbcType::Xyz, simulationState_.box().legacyMatrix, x);
```
Comment (sekelle) [inline src/gromacs/nblib/integrator.cpp]: This is a user-facing function. The fact that there's a legacyMatrix inside box should be hidden from the user
by providing a wrapper.

**Review by sekelle — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/gmxcalculator.h (hunk 15 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxcalculator.h:100]: Done. Made `NbvSetupUtil` a friend of `GmxForceCalculator` like the topology stuff. Once this patch is merged, will turn it into a builder.

File: src/gromacs/nblib/gmxcalculator.h [hunk 19]
```diff
@@ -80,6 +81,8 @@ class GmxForceCalculator
     //! Non-Bonded Verlet object for force calculation
     std::unique_ptr<nonbonded_verlet_t> nbv_;
```
Comment (victorusu) [inline src/gromacs/nblib/gmxcalculator.h]: @kanduri, I think that this object is a manager that encapsulates the NBV object. Therefore, I believe this shouldn't be public, right?

**Review by victorusu — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/gmxcalculator.h (hunk 16 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxcalculator.h]: Done.

File: src/gromacs/nblib/gmxcalculator.h (hunk 19 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxcalculator.h]: Done.

File: src/gromacs/nblib/integrator.cpp (hunk 18 above)
Comment (kanduri) [reply src/gromacs/nblib/integrator.cpp]: The fact that it is a C-array that has to be copied in place means that the solutions are inelegant. Either we expose a member, or return a pointer. 

It is a short-lived interim solution. The moment GROMACS moves to using the new `Matrix` class, this aspect of box will get a lot cleaner.

Comment (ejjordan) [reply src/gromacs/nblib/integrator.cpp]: I don't think this will be such a short-lived solution. Unless you have a plan to refactor all of the PBC code in gromacs to use Matrix, we will have to live with the legacy matrix for quite some time.

Comment (kanduri) [reply src/gromacs/nblib/integrator.cpp]: Done. Turned into a wrapper.

File: src/gromacs/nblib/gmxsetup.h (hunk 17 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: I absolutely get this and that would be the goal of the builder PR. This PR is more to consolidate things within `GmxSetupUtils` by creating functions to carry specific arguments. The members are now private. The public interface would be expanded and used in the forthcoming PR.

File: src/gromacs/nblib/gmxsetup.cpp (hunk 14 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.cpp]: `box_` renamed to `box` to reflect that it is not an internal state anymore. So it is a temporary variable in that scope.

File: src/gromacs/nblib/gmxsetup.h (hunk 17 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: Important design discussion. I get the value of having those as separate functions to be called with specific arguments. And that is exactly how this is shaping up.

However, having code being called in the public interface is probably not necessary. We can have one builder super-function that does the step-by-step building which is essential for debugging, but having those calls live inside `ForceCalculator` is something whose benefits I don't see yet.

```
ForceCalculator::ForceCalculator(const SimulationState& system, const NBKernelOptions& options)
{
    NbvSetupUtil nbvSetupUtil(system, options);

    gmxForceCalculator_ = nbvSetupUtil.setupGmxForceCalculator();
}
```

The above snippet looks super clean to work with. The functions you described can be called inside the `setupGmxForceCalculator`

File: src/gromacs/nblib/box.cpp [hunk 20]
```diff
@@ -61,7 +61,7 @@ Box::Box(real l)
     box_(YY, YY) = l;
     box_(ZZ, ZZ) = l;
 
-    gmx::fillLegacyMatrix(box_, legacyMatrix_);
+    gmx::fillLegacyMatrix(box_, oldMatrix_);
```
Comment (ejjordan) [inline src/gromacs/nblib/box.cpp]: The name oldMatrix seems overly normative and anyway I am curious why you are undertaking these changes that are not asked for while not addressing review comments.

Comment (kanduri) [reply src/gromacs/nblib/box.cpp]: Oops! That was an intermediate change when I was trying to resolve some name conflicts. Corrected.

And I am addressing the review comments. Please take a look at `NbvSetupUtils`. All the functions have been explicitly separated now. If we can merge this one, then the builder PR will begin review as that involves renaming.

File: src/gromacs/nblib/gmxsetup.h (hunk 17 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: @ejjordan Comments/criticism/guidance requested here.

Comment (ejjordan) [reply src/gromacs/nblib/gmxsetup.h]: From the user's perspective, it will not matter how complex or not the builder for GmxForceCalculator is, since it is not a user interactable object. What we want is an interface for the GmxForceCalculatorBuilder that has a similar feel to the TopologyBuilder. What that means is that we want to be able to write a script that one by one adds the pieces to the object that is being built. Something like the following.

`GmxForceCalculatorBuilder fcb;`
`fcb.addMDAtoms(...);`
`fcb.addNonBondedParameters(...);`
`fcb.addInteractionConst(...);`
`// Add more stuff`
`GmxForceCalculator fc = fcb.build();`

Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: I am trying to understand the benefits of making it feel like `TopologyBuilder`.

The `GmxForceCalculator` uses all the data in `SimulationState`. and `NBKernelOptions` even if the `NbvSetupUtil` doesn't. For this reason, the builder must pass both these objects down.

The pieces for this are not scattered the way the data for `Topology` is. The latter needs a variety of molecules constructed in scope.

Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: That analogy is not entirely valid in the case of `NbvSetupUtil` as there isn't much to its own internal state. There is however a multitude of steps in constructing `GmxForceCalculator` and its internal members as you can see in `setupGmxForceCalculator()`.  You can see a semblance of such a script that you described above here.

I do not see benefits yet in making those calls in the constructor of `ForceCalculator`. Please clarify this.

Comment (sekelle) [reply src/gromacs/nblib/gmxsetup.h]: @ejjordan Exactly! @kanduri Yes, the analogy to TopologyBuilder is valid and of course the calculator builder has a lot of internal state, namely all the data that Joe mentioned above.
What Joe outlined above is exactly what we decided to do and I am very confused why this is still unclear or at least, in case you have a different vision, you didn't voice it two weeks ago when we were all in the same room.

Comment (victorusu) [reply src/gromacs/nblib/gmxsetup.h]: @kanduri, Yes! What @ejjordan and @sebkelle1 are saying is what we agreed in the hackathon.

Comment (sekelle) [reply src/gromacs/nblib/gmxsetup.h]: Btw what Joe outlined is the director code. It can be it's own function that's called in the ForceCalculator, but it does neither belong into the gmx calculator builder nor into the gmx calculator itself.

Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: @sebkelle1 @ejjordan 
`NbvSetupUtl` has just two internal members namely `nonbondedParameters_` and `particleInfoAllVdw_` which basically come from `NBKernelOptions`. Unlike the `TopologyBuilder` which takes new molecules, and updates all its internal members based on these inputs. The references to the other two are just stored as they come and there is no logic to set them.

Because there are just two members that basically need the same input, this part is handled within the constructor using dedicated functions so that this process is clearly separated. 

Construction of `GmxForceCalculator` is a multi-stage one and a script for that exists in `setupGmxForceCalculator()`. Since this is meant to be a builder, it certainly belongs there. Each member inside `GmxForceCalculator` is filled individually using dedicated functions wherever possible.

I can do what you asked for if you all really want it, but I don't see the benefits yet of pushing this script inside the constructor of `ForceCalculator`.

Comment (ejjordan) [reply src/gromacs/nblib/gmxsetup.h]: The script does not live in `ForceCalculator` but in an Impl class that is called there.

Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: So do we want another Impl object outside `NbvSetupUtil` that orchestrates the filling of each member of `GmxForceCalculator`?

I am not seeing benefits of doing so yet.

Comment (ejjordan) [reply src/gromacs/nblib/gmxsetup.h]: There are a number of benefits.
1) What is need for object construction is clear since it all happens in one place.
2) As we add functionality that requires more steps in the builder, it is obvious where to add those steps.
3) Steps of building that have lots of logic can drive refactoring efforts of gromacs internals to make the interface layer thinner.
4) Dependencies between steps can be easily managed with asserts in the steps. This can also focus refactoring efforts since it helps show data flow between objects.

Basically having this layer helps keep complexity manageable through expansion of scope and refactoring.

Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: Please check the new push. There is not a `GmxBuildDirector` class introduced.

1. With segregated functions for each field of the `GmxForceCalculator`, that clarification had been introduced.

2. There is a specific function to add those steps both before and now. 

3. Agreed. So having such a script is essential, the discussion was on where that should live. 

4. Good idea. I am drafting a small set of design goals which @sebkelle1 and I spoke about just a while ago.

Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h]: Also, renaming classes would be done after both of the outstanding PRs are merged. Just want to prevent conflicts that emerge solely due to the rename.

File: src/gromacs/nblib/gmxcalculator.h [hunk 21]
```diff
@@ -53,29 +53,33 @@
...
     //! Compute forces and return
     gmx::PaddedHostVector<gmx::RVec> compute();
 
+    //! Puts particles on a grid based on bounds specified by the box
+    void setParticlesOnGrid(std::vector<int>&             particleInfoAllVdw,
+                            const std::vector<gmx::RVec>& coordinates,
+                            const Box&                    box);
+
+private:
+    friend class GmxSetupDirector;
```
Comment (sekelle) [inline src/gromacs/nblib/gmxcalculator.h]: It's okay for the GmxForceCalculatorBuilder (now NbvSetupUtil) to be a friend, but not the director.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 22]
```diff
@@ -215,54 +221,123 @@ std::unique_ptr<nonbonded_verlet_t> NbvSetupUtil::setupNbnxmInstance()
...
-    gmxForceCalculator_p->nbv_ = setupNbnxmInstance();
+std::unique_ptr<GmxForceCalculator> GmxSetupDirector::setupGmxForceCalculator(const SimulationState& system,
+                                                                              const NBKernelOptions& options)
+{
+    NbvSetupUtil nbvSetupUtil;
+    nbvSetupUtil.setExecutionContext(options);
+    nbvSetupUtil.setNonBondedParameters(system.topology());
+    nbvSetupUtil.setParticleInfoAllVdv(system.topology().numParticles());
+
+    auto gmxForceCalculator_p = std::make_unique<GmxForceCalculator>();
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.cpp]: Why not put this pointer into the builder/nbvSetupUtil? Then you also wouldn't have the access the builder's internal guts in the following two lines. It would just be `nbvSetupUtil.setupInteractionConst(options)`, and all would be good.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 23]
```diff
@@ -173,33 +169,36 @@ void NbvSetupUtil::unpackTopologyToGmx()
...
     }
+}
 
-    particleInfoAllVdw_.resize(numParticles);
-    for (size_t particleI = 0; particleI < numParticles; particleI++)
-    {
-        SET_CGINFO_HAS_VDW(particleInfoAllVdw_[particleI]);
-        SET_CGINFO_HAS_Q(particleInfoAllVdw_[particleI]);
-    }
+void NbvSetupUtil::setAtomProperties(const Topology& topology)
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.cpp]: Taking the particle ID and charge vectors here as arguments would be slightly nice I think

File: src/gromacs/nblib/gmxsetup.cpp [hunk 24]
```diff
@@ -173,33 +169,36 @@ void NbvSetupUtil::unpackTopologyToGmx()
...
+    gmxForceCalculator_->mdatoms_.typeA =
+            const_cast<int*>(topology.getParticleTypeIdOfAllParticles().data());
+    gmxForceCalculator_->mdatoms_.chargeA = const_cast<real*>(topology.getCharges().data());
+
+    gmxForceCalculator_->nbv_->setAtomProperties(gmxForceCalculator_->mdatoms_, particleInfoAllVdw_);
 }
 
 //! Sets up and returns a Nbnxm object for the given options and system
-std::unique_ptr<nonbonded_verlet_t> NbvSetupUtil::setupNbnxmInstance()
+void NbvSetupUtil::setupNbnxmInstance(const Topology& topology, const NBKernelOptions& options)
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.cpp]: Here I think only the number of particles is needed instead of the entire topology.

**Review by victorusu — APPROVED** [review]
`lgtm`

File: src/gromacs/nblib/gmxsetup.cpp (hunk 23 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.cpp]: On it. Will change it the next patch,

**Review by sekelle — APPROVED** [review]

File: src/gromacs/nblib/box.cpp [hunk 25]
```diff
@@ -60,6 +60,8 @@ Box::Box(real l)
     box_(XX, XX) = l;
     box_(YY, YY) = l;
     box_(ZZ, ZZ) = l;
+
+    gmx::fillLegacyMatrix(box_, legacyMatrix_);
```
Comment (ejjordan) [inline src/gromacs/nblib/box.cpp:64]: The fact that this is needed in two places highlights a flaw in our implementation. Box(l) should just call box(x, y, z). This can be a followup.

File: src/gromacs/nblib/gmxcalculator.cpp [hunk 26]
```diff
@@ -50,100 +50,37 @@
...
+    const matrix& legacyBox = box.legacyMatrix();
 
-    nbv_->dispatchNonbondedKernel(gmx::InteractionLocality::Local, interactionConst_, stepWork_,
-                                  enbvClearFNo, forcerec_, &enerd_, &nrnb_);
+    GMX_RELEASE_ASSERT(!TRICLINIC(legacyBox), "Only rectangular unit-cells are supported here");
+    const rvec lowerCorner = { 0, 0, 0 };
+    const rvec upperCorner = { legacyBox[XX][XX], legacyBox[YY][YY], legacyBox[ZZ][ZZ] };
 
-    nbv_->atomdata_add_nbat_f_to_f(gmx::AtomLocality::All, verletForces_);
+    const real particleDensity = coordinates.size() / det(legacyBox);
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxcalculator.cpp:79]: This can be a function since it is used more than once.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 27]
```diff
@@ -215,56 +214,128 @@ std::unique_ptr<nonbonded_verlet_t> NbvSetupUtil::setupNbnxmInstance()
...
+    gmxForceCalculator_->verletForces_ =
+            gmx::PaddedHostVector<gmx::RVec>(numParticles, gmx::RVec(0, 0, 0));
+}
 
-    return gmxForceCalculator_p;
+std::unique_ptr<GmxForceCalculator> GmxSetupDirector::setupGmxForceCalculator(const SimulationState& system,
+                                                                              const NBKernelOptions& options)
+{
+    NbvSetupUtil nbvSetupUtil;
+    nbvSetupUtil.setExecutionContext(options);
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxsetup.cpp:323]: Can be options.numThreads in a followup.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 28]
```diff
@@ -215,56 +214,128 @@ std::unique_ptr<nonbonded_verlet_t> NbvSetupUtil::setupNbnxmInstance()
...
+    nbvSetupUtil.setNonBondedParameters(system.topology().getParticleTypes());
+    nbvSetupUtil.setParticleInfoAllVdv(system.topology().numParticles());
+
+    nbvSetupUtil.setupInteractionConst(options);
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxsetup.cpp:327]: Can be setupInteractionConst(options.pairlistCutoff, options.coulombType) in a followup.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 29]
```diff
@@ -215,56 +214,128 @@ std::unique_ptr<nonbonded_verlet_t> NbvSetupUtil::setupNbnxmInstance()
...
+    nbvSetupUtil.setupStepWorkload(options);
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxsetup.cpp:328]: Can be options.computeVirialAndEnergy in a followup.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 30]
```diff
@@ -215,56 +214,128 @@ std::unique_ptr<nonbonded_verlet_t> NbvSetupUtil::setupNbnxmInstance()
...
+void NbvSetupUtil::setParticlesOnGrid(const std::vector<gmx::RVec>& coordinates, const Box& box)
+{
+    gmxForceCalculator_->setParticlesOnGrid(particleInfoAllVdw_, coordinates, box);
+}
 
-    put_atoms_in_box(PbcType::Xyz, box_, system_.coordinates());
+void NbvSetupUtil::constructPairList(const gmx::ListOfLists<int>& exclusions)
+{
+    t_nrnb nrnb;
+    gmxForceCalculator_->nbv_->constructPairlist(gmx::InteractionLocality::Local, exclusions, 0, &nrnb);
```
Comment (ejjordan) [inline src/gromacs/nblib/gmxsetup.cpp:310]: To support DD we will need to be able to pass InteractionLocality. Also, we don't have to use the gmx version, even in our translation layer, since we can just have the same enum order and get the correct int value.

**Review by ejjordan — APPROVED** [review]
